File: Valid Palindrome.cpp
class Solution {
public:
    /**
     * @param s A string
     * @return Whether the string is a valid palindrome
     */
    bool isPalindrome(string& s) {
        // Write your code here
        if (s.empty())
        {
            return true;
        }
        int left = 0, right = s.length() - 1;
        while (left < right)
        {
            while (left < right && !isalnum(s[left]))
            {
                ++left;
            }
            while (right > left && !isalnum(s[right]))
            {
                --right;
            }
            //the worse case: left == right, so
            if (toupper(s[right]) != toupper(s[left]))
            {
                return false;
            }
            else
            {
                ++left;
                --right;
            }
            // if (left <= right)
            // {
            //     if (toupper(s[right]) == toupper(s[left]))
            //     {
            //         --right;
            //         ++left;
            //     }
            //     else
            //     {
            //         return false;
            //     }
            // }

        }
        return true;
    }
};

File: Valid Palindrome.py
class Solution:
    # @param {string} s A string
    # @return {boolean} Whether the string is a valid palindrome
    def isPalindrome(self, s):
        # Write your code here
        
        if not s:
            return True
        length = len(s)
        left, right = 0, length - 1
        while left <= right:
            if not s[left].isalnum():
                left += 1
            elif not s[right].isalnum():
                right -= 1
            else:
                if s[left].lower() != s[right].lower():
                    return False
                else:
                    left += 1
                    right -= 1
        return True